src/PostgREST/App.hs

```diff
@@ -188,21 +188,31 @@ app dbStructure proc conf apiRequest =
                     (iPreferRepresentation apiRequest) []
               row <- H.statement (toS $ pjRaw pJson) stm
               let (_, queryTotal, _, body) = extractQueryResult row
-              if contentType == CTSingularJSON
-                 && queryTotal /= 1
-                 && iPreferRepresentation apiRequest == Full
-                then do
-                  HT.condemn
-                  return $ singularityError (toInteger queryTotal)
-              else do
-                let r = contentRangeH 0 (toInteger $ queryTotal-1)
-                          (toInteger <$> if shouldCount then Just queryTotal else Nothing)
-                    s = if iPreferRepresentation apiRequest == Full
-                          then status200
-                          else status204
-                return $ if iPreferRepresentation apiRequest == Full
-                  then responseLBS s [toHeader contentType, r] (toS body)
-                  else responseLBS s [r] ""
+
+                  updateIsNoOp       = S.null $ pjKeys pJson
+                  contentRangeHeader = contentRangeH 0 (queryTotal - 1) $
+                                          if shouldCount then Just queryTotal else Nothing
+                  minimalHeaders     = [contentRangeHeader]
+                  fullHeaders        = toHeader contentType : minimalHeaders
+
+                  status | queryTotal == 0 && not updateIsNoOp      = status404
+                         | iPreferRepresentation apiRequest == Full = status200
+                         | otherwise                                = status204
+
+              case (contentType, queryTotal, iPreferRepresentation apiRequest) of
+                (CTSingularJSON, 1, Full) ->
+                  return $ responseLBS status fullHeaders (toS body)
+
+                (CTSingularJSON, _, Full) ->
+                  do HT.condemn
+                     return $ singularityError 0
```

Comment:
Previously we had:
```haskell
return $ singularityError (toInteger queryTotal)
```
It'd be better to keep the queryTotal instead of always returning 0.

I'm surprised this wasn't included in any test. Perhaps you could add one for making sure we don't miss this again.

Reply:
Ah, nice catch, thanks!